src/compute/src/logging/compute.rs

```diff
@@ -549,18 +307,331 @@ pub fn construct<A: Allocate>(
+            let elapsed_pow = elapsed_ns.next_power_of_two();
+            let elapsed_ns: i128 = elapsed_ns.try_into().expect("elapsed_ns too big");
+            let diff = (elapsed_ns, 1);
+            self.output.peek_duration.give((elapsed_pow, ts, diff));
+        } else {
+            error!(
+                uuid = ?uuid,
+                "peek not yet registered",
+            );
+        }
+    }
+
+    fn handle_frontier(&mut self, export_id: GlobalId, frontier: Timestamp, diff: i8) {
+        let ts = self.ts();
+        let datum = FrontierDatum {
+            export_id,
+            frontier,
+        };
+        self.output.frontier.give((datum, ts, diff.into()));
+
+        // Everything below only applies to frontier insertions.
+        if diff <= 0 {
+            return;
+        }
+
+        // Check if we have imports associated to this export and report frontier advancement
+        // delays.
+        if let Some(import_map) = self.state.export_imports.get_mut(&export_id) {
+            for (&import_id, delay_state) in import_map {
+                let FrontierDelayState {
+                    time_deque,
+                    delay_map,
+                } = delay_state;
+                while let Some(current_front) = time_deque.pop_front() {
+                    let (import_frontier, update_time) = current_front;
+                    if frontier >= import_frontier {
+                        let elapsed_ns = self.time.as_nanos() - update_time;
+                        let elapsed_pow = elapsed_ns.next_power_of_two();
+                        let elapsed_ns: i128 = elapsed_ns.try_into().expect("elapsed_ns too big");
+                        let (delay_sum, delay_count) = match delay_map.entry(elapsed_pow) {
+                            Entry::Vacant(v) => v.insert((0, 0)),
+                            Entry::Occupied(o) => {
+                                let (sum, count) = o.get().clone();
+                                let datum = FrontierDelayDatum {
+                                    export_id,
+                                    import_id,
+                                    delay_pow: elapsed_pow,
+                                    count,
+                                    sum,
+                                };
+                                self.output.frontier_delay.give((datum, ts, -1));
+                                o.into_mut()
+                            }
+                        };
+                        *delay_sum += elapsed_ns;
+                        *delay_count += 1;
+                        let datum = FrontierDelayDatum {
+                            export_id,
+                            import_id,
+                            delay_pow: elapsed_pow,
+                            count: *delay_count,
+                            sum: *delay_sum,
+                        };
+                        self.output.frontier_delay.give((datum, ts, 1));
```

Comment:
The accumulation of `count` and `sum` could be done in the DD diff field, like it's done for `peek_duration`. We plan to remove the `sum` field as part of #17836, and we'll change the `count` handling as part of that work.